File: src/main/kotlin/com/vlnabatov/alabaster/extensions/annotation/JavaAnnotator.kt
package com.vlnabatov.alabaster.extensions.annotation

import com.intellij.lang.annotation.AnnotationHolder
import com.intellij.lang.annotation.Annotator
import com.intellij.lang.annotation.HighlightSeverity.INFORMATION
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.BRACES
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors.NUMBER
import com.intellij.psi.JavaTokenType.TEXT_BLOCK_LITERAL
import com.intellij.psi.JavaTokenType.TRUE_KEYWORD
import com.intellij.psi.JavaTokenType.FALSE_KEYWORD
import com.intellij.psi.JavaTokenType.NULL_KEYWORD
import com.intellij.psi.JavaTokenType.STRING_LITERAL
import com.intellij.psi.PsiElement
import com.intellij.psi.util.elementType
import com.vlnabatov.alabaster.annotateSeparationMarks


class JavaAnnotator : Annotator {
    override fun annotate(element: PsiElement, holder: AnnotationHolder) {
        try {
            when (element.elementType) {
                TRUE_KEYWORD, FALSE_KEYWORD, NULL_KEYWORD ->
                    holder.newSilentAnnotation(INFORMATION).textAttributes(NUMBER).create()

                STRING_LITERAL -> annotateSeparationMarks(element, holder, BRACES)
                TEXT_BLOCK_LITERAL -> annotateSeparationMarks(element, holder, BRACES, 3)
            }
        } catch (e: Exception) {
            /* Should not happen */
        }
    }
}

File: src/main/clojure/com/vlnabatov/alabaster/java_annotator.clj
(ns com.vlnabatov.alabaster.java-annotator
  (:require [com.vlnabatov.alabaster.util :refer [import-static]]
            [clojure.pprint :as pprint])
  (:import (com.intellij.lang.annotation AnnotationHolder)
           (com.intellij.lang.annotation HighlightSeverity)
           (com.intellij.openapi.editor DefaultLanguageHighlighterColors)
           (com.intellij.psi.tree IElementType)
           (com.vlnabatov.alabaster UtilKt)
           (com.intellij.psi JavaTokenType)
           (com.intellij.psi PsiElement))
  (:gen-class :implements [com.intellij.lang.annotation.Annotator]
              :name       com.vlnabatov.alabaster.extensions.annotation.JavaAnnotator
              :prefix     "-"))


(defn -annotate
  [_ ^PsiElement element ^AnnotationHolder holder]
  (let [^IElementType elementType (.. element getNode getElementType)]
    (cond (#{JavaTokenType/TRUE_KEYWORD JavaTokenType/FALSE_KEYWORD JavaTokenType/NULL_KEYWORD} elementType)
            (do (-> holder
                    (.newSilentAnnotation HighlightSeverity/INFORMATION)
                    (.textAttributes DefaultLanguageHighlighterColors/NUMBER)
                    (.create)))
          (= JavaTokenType/STRING_LITERAL elementType) (do (UtilKt/annotateSeparationMarks element holder))
          (= JavaTokenType/TEXT_BLOCK_LITERAL elementType)
            (do (UtilKt/annotateSeparationMarks element holder DefaultLanguageHighlighterColors/BRACES 3)))))